chalk-solve/src/clauses/generalize.rs

```diff
@@ -0,0 +1,80 @@
+//! This gets rid of free variables in a type by replacing them by fresh bound
+//! ones. We use this when building clauses that contain types passed to
+//! `program_clauses`; these may contain variables, and just copying those
+//! variables verbatim leads to problems. Instead, we return a slightly more
+//! general program clause, with new variables in those places.
```

Comment:
Hmm, ok, is this specific to `dyn` types? I think all the other program clauses are generated purely from declaration sites, but I guess those do kind of "mix" data from the goal, and I can see how that could lead to problems. It'd be good to give some examples.

Reply:
Yes, it's just `dyn` at the moment

Reply:
Yeah I can't think of another case where it would arise -- it would basically be any case where we are generating rules *from the goal itself* and not from definitions that the goal references, but I think that's specific to `dyn` right now.